mathematica/retrobop_dynamical_stability.cpp

```diff
@@ -0,0 +1,668 @@
+            extremal_separations_in_m[moon].back() ==
+                record_separation_in_m[moon]) {
+          LOG(WARNING) << "After "
+                       << times_in_s[moon].back() * Second / JulianYear
+                       << " a, " << names[moon] << " has an apsis at "
+                       << extremal_separations_in_m[moon].back() * Metre;
+        }
+      }
+      last_separations[moon] = separation;
+      last_separation_changes.at(moon) = separation_change;
+    }
+
+    tylo_bop_separations_in_m.emplace_back(
+        (EvaluatePosition(ephemeris, Tylo, t) -
+         EvaluatePosition(ephemeris, Bop, t)).Norm() / Metre);
+    pol_bop_separations_in_m.emplace_back(
+        (EvaluatePosition(ephemeris, Pol, t) -
+         EvaluatePosition(ephemeris, Bop, t)).Norm() / Metre);
+
+    {
+      // KSP's osculating elements.
+      auto const bop_elements =
+          KeplerOrbit<Barycentric>(
+              *ephemeris.bodies()[Jool],
+              MasslessBody(),
+              EvaluateDegreesOfFreedom(ephemeris, Bop, t) -
+                  EvaluateDegreesOfFreedom(ephemeris, Jool, t), t)
+              .elements_at_epoch();
+      bop_eccentricities.emplace_back(bop_elements.eccentricity);
+      bop_inclinations_in_degrees.emplace_back(bop_elements.inclination /
+                                               Degree);
+      bop_nodes_in_degrees.emplace_back(
+          bop_elements.longitude_of_ascending_node / Degree);
+      bop_arguments_of_periapsis_in_degrees.emplace_back(
+          bop_elements.argument_of_periapsis / Degree);
+    }
+
+    {
+      BarycentreCalculator<DegreesOfFreedom<Barycentric>,
+                           GravitationalParameter>
+          innermost_jool_system;
+      for (auto const celestial : {Jool, Laythe, Vall, Tylo}) {
+        innermost_jool_system.Add(
+            EvaluateDegreesOfFreedom(ephemeris, celestial, t),
+            ephemeris.bodies()[celestial]->gravitational_parameter());
+      }
+      auto const bop_jacobi_elements =
+          KeplerOrbit<Barycentric>(MassiveBody(innermost_jool_system.weight()),
+                                   *ephemeris.bodies()[Bop],
+                                   EvaluateDegreesOfFreedom(ephemeris, Bop, t) -
+                                       innermost_jool_system.Get(),
+                                   t).elements_at_epoch();
+      bop_jacobi_eccentricities.emplace_back(bop_jacobi_elements.eccentricity);
+      bop_jacobi_inclinations_in_degrees.emplace_back(
+          bop_jacobi_elements.inclination / Degree);
+      bop_jacobi_nodes_in_degrees.emplace_back(
+          bop_jacobi_elements.longitude_of_ascending_node / Degree);
+      bop_jacobi_arguments_of_periapsis_in_degrees.emplace_back(
+          bop_jacobi_elements.argument_of_periapsis / Degree);
+    }
+  }
+
+  std::ofstream file;
+  file.open("retrobop_century.generated.wl");
+  file << mathematica::Assign("laytheTimes", times_in_s[Laythe]);
```

Comment:
I would prefer to keep the strongly-typed quantities up to this point (that would avoid the ugly multiplications/divisions by `Metre` and friends) and have a small function to remove typing here:
```
template<typename Q>
std::vector<double> RemoveTheTypesDamnit(std::vector<Q> const& q) {
  for ... {
    result.push_back(q[i] / SIUnit<Q>());
  }
}
```